lib/spack/spack/environment.py

```diff
@@ -1703,7 +1703,22 @@ def matching_spec(self, spec):
         # Dependency-only specs will have value None
         matches = {}
 
+        if not isinstance(spec, spack.spec.Spec):
+            spec = spack.spec.Spec(spec)
+
         for user_spec, concretized_user_spec in self.concretized_specs():
+            # Deal with concrete specs differently
+            if spec.concrete:
+                # Matching a concrete spec is more restrictive
+                # than just matching the dag hash
+                is_match = (
+                    spec in concretized_user_spec and
+                    concretized_user_spec[spec.name].build_hash() == spec.build_hash()
+                )
+                if is_match:
+                    matches[spec] = spec
+                continue
+
```

Comment:
Can this fix be moved to a different PR since it applies for Cloud CI whether `clingo` is the default or explicitly requested by the user?

I'd also hate to see this fix delayed since I've run into the problem for at least one, maybe two different CI efforts.